### PR TITLE
:memo: Improve variables docs

### LINKED ISSUE #1976: Update variables doc
### Where on docs.platform.sh should be changed?

https://docs.platform.sh/development/variables.html

### What exactly should be updated?

Incorporate ideas from #1966

Update for:
- Content accuracy
- Includes what it should
- Doesn't include what doesn't fit
- Fits template
- Fits style guide

### Additional context

Probably split how-to content (like how to access) from reference content.

Make sure different scopes (build, runtime) are clear.

### PR DESCRIPTION
**Pull request by CollierCZ**
## Why

Closes #1976

## What's changed

* Got rid of https://docs.platform.sh/bestpractices/environment-build.html and moved relevant content elsewhere
* Created separate pages for setting and reusing variables
* Updated variables content for clarity and style
* Fixed a couple of Vale bugs

### CODE REVIEW

File: docs/src/development/variables/_index.md [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+title: Variables overview
+weight: 5
+description: |
+  Variables give you control over your project's build process and runtime environment. You can set them in your code to make changes across your project or independent of the code for environment-specific settings.
+---
+
+{{< description >}}
+
+In this way, your app has additional information, such as database credentials, the host or port it can use, and which server to connect to.
+
```
[the remark this answers is not in the document]
Comment (CollierCZ) [reply docs/src/development/variables/_index.md:11]: I moved the table of the types of variables to the overview and tried to include more examples. See if that helps?

File: docs/src/development/variables/_index.md [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+<!-- vale Vale.Spelling = YES -->
+
+Project variables that conflict with environment variables are labeled as **Inactive**.
+Environment variables are labeled as **Inherited** when they get their value from a parent environment
+and as **Overridden** when there is a conflict and the parent environment's value doesn't apply.
+
+## Variable prefixes
+
+Certain variable name prefixes have special meanings.
+Some are defined by Platform.sh and apply automatically.
```
Comment (trolologuy) [inline docs/src/development/variables/_index.md:142]:
```suggestion
Some are defined by Platform.sh and apply automatically ( such as `PLATFORM_<name of variable>`).
```

Comment (CollierCZ) [reply docs/src/development/variables/_index.md:142]: Those ones are treated as a separate case and aren't what's meant here. It's prefixes like `env` here that apply. If you add `PLATFORM_` as a prefix, I don't think it does anything special.

File: docs/src/development/variables/_index.md [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+Any variable with the prefix `php:` is added to the `php.ini` configuration for all PHP-based application containers in the project.
+
+For example, an environment variable named `php:display_errors` with the value `On` is equivalent to placing the following in `php.ini`:
+
+```ini
+display_errors = On
+```
+
+This feature is primarily useful to override debug configuration on development environments,
+such as enabling errors and configuring the Xdebug extension.
```
Comment (trolologuy) [inline docs/src/development/variables/_index.md:175]:
```suggestion
such as enabling errors and [configuring the Xdebug extension](../../languages/php/xdebug.md).
```

File: docs/src/development/variables/_index.md [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+
+To apply a setting to all environments or have a setting differ among multiple PHP containers in one project,
+specify the variables in the `.platform.app.yaml` file for your application.
+See the [PHP configuration page](../../languages/php/ini.md) for more information.
+
+### Framework-specific variables
+
+For specific frameworks, you can implement logic to override global configurations with environment-specific variables.
+
+The Drupal templates show examples of logic mapping variables to Drupal's configuration system.
```
Comment (trolologuy) [inline docs/src/development/variables/_index.md:185]:
```suggestion
The [Drupal templates](https://github.com/platformsh-templates/?q=drupal) show examples of logic mapping variables to Drupal's configuration system.
```

File: docs/src/development/variables/_index.md [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+| -------------------------------------------------- | ----------- | ----------- | ---------- | ----- | -------- |----- |
+| [Application](./set-variables.md#set-variables-in-your-app) | Application | Application | 4          | Yes   | Yes      | Non-secret values that are the same across all environments |
+| [Project](./set-variables.md#create-project-variables)               | User        | Project     | 3          | Yes   | Yes      | Secret values that are the same across all environments, such as database credentials |
+| [Environment](./set-variables.md#create-environment-specific-variables)       | User        | Environment | 2          | Some  | Yes      | Values that vary by environment, such as which database to connect to or which payment API keys to use |
+| [Platform.sh](./use-variables.md#use-platformsh-provided-variables)  | Pre-defined | Environment | 1          | Some  | Yes      | For information about your Platform.sh project |
+
+If there are conflicts between variables with the same name, variables [take precedence](#overrides) from 1 down.
+So Platform.sh-provided values override environment variables, which override project variables,
```
Comment (trolologuy) [inline docs/src/development/variables/_index.md:27]:
```suggestion
So Platform.sh-provided values override environment variables (4), which override project variables (3),
```

File: docs/src/development/variables/set-variables.md [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+platform variable:create --level project --name foo --value bar --visible-build true --visible-runtime false
+```
+
+You can also change the variable options after you create them (except for sensitive values, which can't be set to non-sensitive). For example, to make the `foo` variable visible at runtime and hidden during the build, run this command:
+
+```bash
+platform variable:update foo --visible-build false --visible-runtime true
+```
+
+Note that any changes to project variables require you to deploy your environments to have effect.
```
Comment (trolologuy) [inline docs/src/development/variables/set-variables.md:61]:
```suggestion
Note that any changes to project variables require you to (re)deploy your environments to have effect.
```

File: docs/src/development/variables/set-variables.md [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+
+```bash
+platform variable:create -e main --name paypal_id --inheritable false --sensitive true
+```
+
+Other environments don't inherit it and get either a project variable of the same name if it exists or no value at all.
+
+Note that changing an environment variable causes that environment to be redeployed so the new value is available.
+However, child environments are *not* redeployed.
+To make the new value accessible to those environments, redeploy them manually.
```
Comment (trolologuy) [inline docs/src/development/variables/set-variables.md:97]:
```suggestion
To make the new value accessible to those environments, redeploy them either via:
- [the console](../../administration/web/_index.md),
- [the cli](/cli.md) `platform redeploy -e <name of the child environment>`.
```

File: docs/src/development/variables/set-variables.md [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+
+### Example environment variable
+
+Environment variables are a good place to store values that apply only on Platform.sh and not on your local development environment. This includes API credentials for third-party services, mode settings, and which server (development vs. production) to use.
+
+One example would be to define a Node.js application's build on a production branch (`NODE_ENV=production`),
+but use development mode (`NODE_ENV=development`) for each of your development environments.
+Assuming you have a `main` environment for production and then a `staging` environment with more child environments for development,
```
Comment (trolologuy) [inline docs/src/development/variables/set-variables.md:105]:
```suggestion
Assuming you have a `main` environment for production and a `staging` environment with more child environments for development,
```

File: docs/src/development/variables/set-variables.md [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+```
+
+Note that the file is sourced after all other environment variables above are defined and so they're available to the script.
+This also means the `.environment` script has the last word on environment variable values and can override anything it wants to.
+
+## Map variables
+
+If your app needs different names for environment variable than those set by Platform.sh, which is common for database connections,
+map the Platform.sh's variable names to those required by the application.
+Do this in the app with the help of the Config Reader libraries or via a shell script.
```
Comment (trolologuy) [inline docs/src/development/variables/set-variables.md:138]:
```suggestion
Do this in the app with the help of the [Config Reader libraries](https://github.com/platformsh?q=config-reader), ) or via a shell script.
```

File: docs/src/development/variables/_index.md (hunk 4 above)
Comment (CollierCZ) [reply docs/src/development/variables/_index.md:185]: The specific relevant templates are linked right below so I don't think the more general link is necessary here.

File: docs/src/development/variables/set-variables.md [hunk 10]
```diff
@@ -0,0 +1,158 @@
...
+## Set variables in your app
+
+Set variables [in code](../../configuration/app/app-reference.md#variables) using the `.platform.app.yaml` file.
+These values are the same across all environments and present in the Git repository,
+which makes them a poor fit for API keys and other such secrets.
+
+They're better fits for uses such as configuration for consistent builds across every environment,
+including setting [PHP configuration values](./_index.md#php-specific-variables).
+
+Application variables are available at both build time and runtime.
```
[the remark this answers is not in the document]
Comment (CollierCZ) [reply docs/src/development/variables/set-variables.md:19]: That page is a description of the overall process. I feel like knowing what build time is is a prerequisite for doing things like variables. I'm not sure how much that specific page helps with the understanding of this step in particular. Maybe you could explain what you hope they get out of it?

File: docs/src/development/variables/set-variables.md (hunk 7 above)
Comment (CollierCZ) [reply docs/src/development/variables/set-variables.md:97]: We also already have instructions for this: https://docs.platform.sh/development/troubleshoot.html#force-a-redeploy
So I'd just link there.

File: docs/src/development/variables/set-variables.md (hunk 9 above)
Comment (CollierCZ) [reply docs/src/development/variables/set-variables.md:138]: I thought I'd included this :facepalm:

File: docs/src/development/variables/set-variables.md [hunk 11]
```diff
@@ -0,0 +1,158 @@
...
+It uses the `jq` library, which is included in all app containers for this purpose.
+
+```bash {location=".environment"}
+export APP_DATABASE_HOST=$(echo $PLATFORM_RELATIONSHIPS | base64 --decode | jq -r ".database[0].host")
+export APP_DATABASE_PORT=$(echo $PLATFORM_RELATIONSHIPS | base64 --decode | jq -r ".database[0].port")
+export APP_DATABASE_NAME=$(echo $PLATFORM_RELATIONSHIPS | base64 --decode | jq -r ".database[0].path")
+export APP_DATABASE_USER=$(echo $PLATFORM_RELATIONSHIPS | base64 --decode | jq -r ".database[0].username")
+export APP_DATABASE_PASSWORD=$(echo $PLATFORM_RELATIONSHIPS | base64 --decode | jq -r ".database[0].password")
+```
+
```
Comment (trolologuy) [inline docs/src/development/variables/set-variables.md:151]:
```suggestion
To use it:
1. Add the `.environment` file to your project and adapt the exported variables if needed.
1. [Set the environment variables](#create-environment-specific-variables) (`APP_DATABASE_HOST`, `APP_DATABASE_PORT`, ...).

Now, the values can be accessed in the `env` directly.
```

Comment (trolologuy) [reply docs/src/development/variables/set-variables.md:151]: Or is my understanding of the feature, wrong?

File: docs/src/development/variables/set-variables.md [hunk 12]
```diff
@@ -0,0 +1,158 @@
...
+Do this in the app with the help of the Config Reader libraries or via a shell script.
+
+For example, the following [`.environment` script](#set-variables-via-script) exports variables that are visible to the application.
+It uses the `jq` library, which is included in all app containers for this purpose.
+
+```bash {location=".environment"}
+export APP_DATABASE_HOST=$(echo $PLATFORM_RELATIONSHIPS | base64 --decode | jq -r ".database[0].host")
+export APP_DATABASE_PORT=$(echo $PLATFORM_RELATIONSHIPS | base64 --decode | jq -r ".database[0].port")
+export APP_DATABASE_NAME=$(echo $PLATFORM_RELATIONSHIPS | base64 --decode | jq -r ".database[0].path")
+export APP_DATABASE_USER=$(echo $PLATFORM_RELATIONSHIPS | base64 --decode | jq -r ".database[0].username")
```
Comment (trolologuy) [inline docs/src/development/variables/set-variables.md:148]:
```suggestion
export APP_DATABASE_USERNAME=$(echo $PLATFORM_RELATIONSHIPS | base64 --decode | jq -r ".database[0].username")
```
same as above

File: docs/src/development/variables/_index.md (hunk 2 above)
Comment (trolologuy) [reply docs/src/development/variables/_index.md:142]: ah, indeed that makes more sense!

File: docs/src/development/variables/set-variables.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (trolologuy) [reply docs/src/development/variables/set-variables.md:19]: The idea was to help people find the high-level explanation of concepts we bring. Indeed it is a prerequisite, but I thought to point to the right spot in the documentation could help make sure the understanding from the customer perspective is correct. Kind of a reminder?

File: docs/src/development/variables/set-variables.md (hunk 7 above)
Comment (trolologuy) [reply docs/src/development/variables/set-variables.md]: Edited the suggestion to link to the page you provided ;) thanks!

File: docs/src/development/variables/set-variables.md (hunk 11 above)
Comment (CollierCZ) [reply docs/src/development/variables/set-variables.md:151]: The `.environment` file should set the variables as environment variables on its own. That's what it's there for, as far as I know.

**Review by trolologuy — APPROVED** [review]
- [x] The final built page is working as expected.
- [x] The information is correct.
- [x] The information is comprehensive.
- [x] The information is concise.
- [x] The information is clear.

**Review by trolologuy — APPROVED** [review]

**Review by trolologuy — APPROVED** [review]

**Pull request merged**